dm/master/openapi.go

```diff
@@ -0,0 +1,152 @@
+	// middlewares
+	e.Use(docMW)
+	// set logger
+	logger := log.L().Logger
+	logger = logger.With(zap.String("component", "openapi"))
+	e.Use(openapi.ZapLogger(logger))
+	e.Use(echomiddleware.Recover())
+	// disables swagger server name validation. it seems to work poorly
+	swagger.Servers = nil
+	// use our validation middleware to check all requests against the OpenAPI schema.
+	e.Use(middleware.OapiRequestValidator(swagger))
+	openapi.RegisterHandlers(e, s)
+	s.echo = e
+}
+
+// redirectRequestToLeader is used to redirect the request to leader.
+// because the leader has some data in memory, only the leader can process the request.
+func (s *Server) redirectRequestToLeader(ctx context.Context) (needRedirect bool, host string, err error) {
+	isLeader, _ := s.isLeaderAndNeedForward(ctx)
+	if isLeader {
+		return false, s.cfg.AdvertiseAddr, nil
+	}
+	// nolint:dogsled
+	_, _, leaderOpenAPIAddr, err := s.election.LeaderInfo(ctx)
+	return true, leaderOpenAPIAddr, err
+}
+
+// DMAPICreateSource url is:(POST /api/v1/sources).
+func (s *Server) DMAPICreateSource(ctx echo.Context) error {
+	return nil
+}
+
+// DMAPIGetSourceList url is:(GET /api/v1/sources).
+func (s *Server) DMAPIGetSourceList(ctx echo.Context) error {
+	return nil
+}
+
+// DMAPIDeleteSource url is:(DELETE /api/v1/sources).
+func (s *Server) DMAPIDeleteSource(ctx echo.Context, sourceName string) error {
+	return nil
+}
+
+// DMAPIStartRelay url is:(POST /api/v1/sources/{source-id}/relay).
+func (s *Server) DMAPIStartRelay(ctx echo.Context, sourceName string) error {
+	return nil
+}
+
+// DMAPIStopRelay url is:(DELETE /api/v1/sources/{source-id}/relay).
+func (s *Server) DMAPIStopRelay(ctx echo.Context, sourceName string) error {
+	return nil
+}
+
+// DMAPIGetSourceStatus url is:(GET /api/v1/sources/{source-id}/status).
+func (s *Server) DMAPIGetSourceStatus(ctx echo.Context, sourceName string) error {
+	return nil
+}
+
+// DMAPIStartTask url is:(POST /api/v1/tasks).
+func (s *Server) DMAPIStartTask(ctx echo.Context) error {
+	return nil
+}
+
+// DMAPIDeleteTask url is:(DELETE /api/v1/tasks).
+func (s *Server) DMAPIDeleteTask(ctx echo.Context, taskName string) error {
+	return nil
+}
+
+// DMAPIGetTaskList url is:(GET /api/v1/tasks).
+func (s *Server) DMAPIGetTaskList(ctx echo.Context) error {
+	return nil
+}
+
+// DMAPIGetTaskStatus url is:(GET /api/v1/tasks/{task-name}/status).
+func (s *Server) DMAPIGetTaskStatus(ctx echo.Context, taskName string) error {
+	return nil
+}
+
+func terrorHTTPErrorHandler(err error, c echo.Context) {
+	var code int
+	var msg string
+	if tErr, ok := err.(*terror.Error); ok {
+		code = int(tErr.Code())
+		msg = tErr.Error()
+	} else {
+		msg = err.Error()
+	}
+	if sendErr := sendHTTPErrorResp(c, code, msg); sendErr != nil {
+		c.Logger().Error(sendErr)
+	}
+}
+
+func sendHTTPErrorResp(ctx echo.Context, code int, message string) error {
+	err := openapi.ErrorWithMessage{ErrorMsg: message, ErrorCode: code}
+	return ctx.JSON(http.StatusBadRequest, err)
```

Comment:
maybe we should use 5xx because this is server side error?

Reply:
Most cases here return an error from a failed request parameter check, so I think 4xx is more appropriat
5xx will be returned automatically when the program painc